readthedocs/projects/views/public.py

```diff
@@ -266,71 +267,50 @@ def project_downloads(request, project_slug):
     )
 
 
-def project_download_media(request, project_slug, type_, version_slug):
-    """
-    Download a specific piece of media.
+class ProjectDownloadMedia(ServeDocsMixin, View):
 
-    Perform an auth check if serving in private mode.
+    def get(self, request, project_slug, type_, version_slug):
 
-    .. warning:: This is linked directly from the HTML pages.
-                 It should only care about the Version permissions,
-                 not the actual Project permissions.
-    """
-    version = get_object_or_404(
-        Version.objects.public(user=request.user),
-        project__slug=project_slug,
-        slug=version_slug,
-    )
+        """
+        Download a specific piece of media.
 
-    # Send media download to analytics - sensitive data is anonymized
-    analytics_event.delay(
-        event_category='Build Media',
-        event_action=f'Download {type_}',
-        event_label=str(version),
-        ua=request.META.get('HTTP_USER_AGENT'),
-        uip=get_client_ip(request),
-    )
+        Perform an auth check if serving in private mode.
 
-    if settings.DEFAULT_PRIVACY_LEVEL == 'public' or settings.DEBUG:
+        .. warning:: This is linked directly from the HTML pages.
+                     It should only care about the Version permissions,
+                     not the actual Project permissions.
+        """
+        version = get_object_or_404(
+            Version.objects.public(user=request.user),
+            project__slug=project_slug,
+            slug=version_slug,
+        )
+
+        # Send media download to analytics - sensitive data is anonymized
+        analytics_event.delay(
+            event_category='Build Media',
+            event_action=f'Download {type_}',
+            event_label=str(version),
+            ua=request.META.get('HTTP_USER_AGENT'),
+            uip=get_client_ip(request),
+        )
 
         storage = get_storage_class(settings.RTD_BUILD_MEDIA_STORAGE)()
         storage_path = version.project.get_storage_path(
             type_=type_, version_slug=version_slug,
             version_type=version.type,
         )
-        if storage.exists(storage_path):
-            return HttpResponseRedirect(storage.url(storage_path))
 
-        media_path = os.path.join(
-            settings.MEDIA_URL,
-            type_,
-            project_slug,
-            version_slug,
-            '%s.%s' % (project_slug, type_.replace('htmlzip', 'zip')),
+        # URL without scheme and domain to perform an NGINX internal redirect
+        url = storage.url(storage_path)
+        url = urlparse(url)._replace(scheme='', netloc='').geturl()
+
+        return self._serve_media_nginx(
```

Comment:
I think we probably want to just add a `media` flag to `_serve_docs`, and then it can handle the filename logic. I think we still want it to serve via Python when we have `PYTHON_MEDIA` on, so we should really be calling `_serve_docs` at the top-level, instead of only nginx.

Reply:
I added a `download` argument to `_serve*` views. I needed to pass the `version_slug` all around these views to use them when creating the filename :(